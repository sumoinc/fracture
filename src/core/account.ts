import { Component } from "projen";
import { Organization } from "./organization";
import { OrganizationalUnit } from "./organizational-unit";
import { Region, RegionOptions } from "./region";

export interface AccountOptions {
  id: string;
  name?: string;
  rootEmail?: string;
  isManagementAccount?: boolean;
  organizationalUnit?: OrganizationalUnit;
}

export class Account extends Component {
  // member components
  public readonly regions: Region[];
  // parent
  public readonly organization: Organization;
  // all other options
  public readonly options: AccountOptions;

  constructor(organization: Organization, options: AccountOptions) {
    super(organization.project);

    /***************************************************************************
     *
     * INIT OPERATION
     *
     **************************************************************************/

    // member components
    this.regions = [];

    // parents + inverse
    this.organization = organization;
    this.organization.accounts.push(this);

    // all other options
    const defaultOptions: Partial<AccountOptions> = {
      isManagementAccount: false,
    };
    this.options = { ...defaultOptions, ...options };

    // if OU defined, add account to OU
    if (this.options.organizationalUnit) {
      this.options.organizationalUnit.accounts.push(this);
    }

    // debugging output
    this.project.logger.info(`INIT Account: "${this.id}"`);
  }

  public get id() {
    return this.options.id;
  }

  public get name() {
    return this.options.name;
  }

  public get rootEmail() {
    return this.options.rootEmail;
  }

  public get isManagementAccount() {
    return this.options.isManagementAccount;
  }

  public get organizationalUnit() {
    return this.options.organizationalUnit;
  }

  /*****************************************************************************
   *
   *  Configuration Helpers
   *
   ****************************************************************************/

  /**
   * Add an region to an account.
   *
   * @param {RegionOptions}
   * @returns {Region}
   */
  public addRegion(options: RegionOptions) {
    const region = this.regions.find((r) => r.id === options.id);
    return region ? region : new Region(this, options);
  }
}
