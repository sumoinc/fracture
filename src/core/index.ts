export * from "./access-pattern";
export * from "./account";
export * from "./audit-strategy";
export * from "./fracture";
export * from "./index";
export * from "./naming-strategy";
export * from "./operation";
export * from "./organization";
export * from "./organizational-unit";
export * from "./region";
export * from "./resource-attribute";
export * from "./resource";
export * from "./service";
export * from "./structure";
