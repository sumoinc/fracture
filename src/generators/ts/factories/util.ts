import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
