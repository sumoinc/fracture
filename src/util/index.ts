export * from "./determineBuildId";
export * from "./test-util";
