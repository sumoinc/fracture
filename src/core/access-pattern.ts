import { deepMerge } from "projen/lib/util";
import { ValueOf } from "type-fest";
import { Operation, OPERATION_SUB_TYPE } from "./operation";
import { Resource } from "./resource";
import {
  ResourceAttribute,
  ResourceAttributeGenerator,
  ResourceAttributeOptions,
} from "./resource-attribute";
import { FractureComponent } from "../core/component";
import { DynamoGsi } from "../dynamodb/dynamo-gsi";
import { DynamoTable } from "../dynamodb/dynamo-table";

export const ACCESS_PATTERN_TYPE = {
  /**
   * Main PK and PK pattern used to identify this record.
   */
  IDENTIFIER: "Identifier",
  /**
   * Lookup pattern for quick begins_with() searching and such.
   */
  LOOKUP: "lookup",
  /**
   * Any other AP
   */
  OTHER: "other",
} as const;

export interface AccessPatternOptions {
  dynamoGsi?: DynamoGsi;
  pkAttributeOptions?: ResourceAttributeOptions;
  skAttributeOptions?: ResourceAttributeOptions;
  type?: ValueOf<typeof ACCESS_PATTERN_TYPE>;
}

export class AccessPattern extends FractureComponent {
  // member components
  public pkAttribute: ResourceAttribute;
  public skAttribute: ResourceAttribute;
  public operations: Operation[] = [];
  // parent
  public readonly resource: Resource;
  // all other options
  public readonly options: Required<AccessPatternOptions>;

  constructor(resource: Resource, options: AccessPatternOptions) {
    super(resource.fracture);

    /***************************************************************************
     *
     * DEFAULT OPTIONS
     *
     **************************************************************************/

    // make a new gsi is one wasn't passed into the Access Pattern
    const dynamoGsi = options.dynamoGsi
      ? options.dynamoGsi
      : new DynamoGsi(resource.dynamoTable);

    const defaultOptions: Required<AccessPatternOptions> = {
      dynamoGsi,
      pkAttributeOptions: {
        name: dynamoGsi.pkName,
        isPublic: false,
        generator: ResourceAttributeGenerator.COMPOSITION,
        generateOn: [
          OPERATION_SUB_TYPE.CREATE_ONE,
          OPERATION_SUB_TYPE.READ_ONE,
          OPERATION_SUB_TYPE.UPDATE_ONE,
          OPERATION_SUB_TYPE.DELETE_ONE,
        ],
      },
      skAttributeOptions: {
        name: dynamoGsi.skName,
        isPublic: false,
        generator: ResourceAttributeGenerator.COMPOSITION,
        generateOn: [
          OPERATION_SUB_TYPE.CREATE_ONE,
          OPERATION_SUB_TYPE.READ_ONE,
          OPERATION_SUB_TYPE.UPDATE_ONE,
          OPERATION_SUB_TYPE.DELETE_ONE,
        ],
      },
      type: ACCESS_PATTERN_TYPE.OTHER,
    };

    /***************************************************************************
     *
     * INIT ACCESS PATTERN
     *
     **************************************************************************/

    // all other options
    this.options = deepMerge([
      defaultOptions,
      options,
    ]) as Required<AccessPatternOptions>;
    // the attribute might already be defined.
    const pkAttribute = resource.getAttributeByName(dynamoGsi.pkName);
    const skAttribute = resource.getAttributeByName(dynamoGsi.skName);

    // member components
    this.pkAttribute = pkAttribute
      ? pkAttribute
      : new ResourceAttribute(resource, this.options.pkAttributeOptions);
    this.skAttribute = skAttribute
      ? skAttribute
      : new ResourceAttribute(resource, this.options.skAttributeOptions);

    // parent + inverse
    this.resource = resource;
    this.resource.accessPatterns.push(this);

    this.project.logger.info(
      `Access Pattern for: "${this.pkAttribute.name}:${this.skAttribute.name}" initialized.`
    );

    return this;
  }

  public build() {
    this.project.logger.debug(
      `BUILD Access Pattern: "${this.pkAttribute.name}:${this.skAttribute.name}" called.`
    );
    /***************************************************************************
     *
     * CRUD OPERATIONS
     *
     * TODO
     *
     **************************************************************************/
    /*
    new Operation(this, {
      operationType: OPERATION_TYPE.MUTATION,
      operationSubType: OPERATION_SUB_TYPE.LIST,
    });
    */
  }

  addPkAttributeSource(sourceAttribute: ResourceAttribute) {
    this.pkAttribute.compositionSources.push(sourceAttribute);
  }

  addSkAttributeSource(sourceAttribute: ResourceAttribute) {
    this.skAttribute.compositionSources.push(sourceAttribute);
  }

  /**
   * Returns operation for given sub-type
   */
  public getOperation(
    operationSubType: ValueOf<typeof OPERATION_SUB_TYPE>
  ): Operation {
    const returnOperation = this.operations.find(
      (o) => o.operationSubType === operationSubType
    );

    if (!returnOperation) {
      throw new Error(`Operation not found for sub-type: ${operationSubType}`);
    }

    return returnOperation;
  }

  public get createOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.CREATE_ONE);
  }

  public get readOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.READ_ONE);
  }

  public get updateOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.UPDATE_ONE);
  }

  public get deleteOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.DELETE_ONE);
  }

  public get isKeyAccessPattern(): boolean {
    return this.dynamoTable.keyDynamoGsi === this.dynamoGsi;
  }

  public get isLookupAccessPattern(): boolean {
    return this.dynamoTable.lookupDynamoGsi === this.dynamoGsi;
  }

  public get dynamoTable(): DynamoTable {
    return this.dynamoGsi.dynamoTable;
  }

  public get dynamoGsi(): DynamoGsi {
    return this.options.dynamoGsi;
  }
}
