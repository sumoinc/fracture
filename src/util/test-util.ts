import { Project } from "projen";
import { synthSnapshot } from "projen/lib/util/synth";

export const TEST_ACCOUNT_ONE = "000000000000";
export const TEST_ORG_ONE = "org-123456";
export const TEST_REGION_ONE = "us-east-1";

/**
 *
 * Helper to generate and return one file from fracture.
 * Useful when trying to test the final output of generated files.
 *
 * @param fracture
 * @param filepath
 * @returns collection of files starting with supplied filepath
 */
export const synthFiles = (project: Project, filepath: string = ""): any => {
  const snapshot = synthSnapshot(project);
  // console.log(Object.keys(snapshot));
  const filtered = Object.keys(snapshot)
    .filter((path) => path.startsWith(filepath))
    .reduce((obj, key) => {
      return {
        ...obj,
        [key]: snapshot[key],
      };
    }, {} as { [key: string]: any });
  return filtered;
};

export const synthFile = (project: Project, filepath: string): string => {
  const files = synthFiles(project, filepath);

  if (files.length > 1) {
    throw new Error(`more than one file matched filepath "${filepath}"`);
  }

  if (files.length === 0) {
    throw new Error(`No files matched filepath "${filepath}"`);
  }

  return files[filepath];
};

// /**
//  * Builds a simple typescript app project as a test harness
//  */
// export const testProject = () => {
//   return new Project({
//     name: "my-project",
//   });
// };

// export const testNodeProject = (options: Partial<NodeProjectOptions> = {}) => {
//   return new NodeProject({
//     name: "my-project",
//     defaultReleaseBranch: "main",
//     packageManager: NodePackageManager.PNPM,
//     pnpmVersion: "8",
//     ...options,
//   });
// };

// /**
//  * Builds a simple typescript app project as a test harness
//  */
// export const testFractureProject = (
//   options: Partial<FractureProjectOptions> = {}
// ) => {
//   const fp = new FractureProject({
//     name: "my-project",
//     ...options,
//   });

//   /*
//   fp.buildEnvironment = new AwsEnvironment(fp, {
//     name: "test",
//     account: TEST_ACCOUNT_ONE,
//     region: TEST_REGION_ONE,
//     primaryHostedZoneId: "foo",
//     primaryDomainName: "bar",
//   });
//   */

//   return fp;
// };

// /**
//  * Builds a simple data service as a test harness
//  */
// export const testApp = () => {
//   return new App({
//     parent: testFractureProject(),
//     name: "my-app",
//   });
// };

// /**
//  * Builds a simple data service as a test harness
//  */
// export const testDataService = () => {
//   return new DataService(testApp(), {
//     name: "my-service",
//   });
// };

// /**
//  * Build a simple resource to test with
//  */
// export const testResource = () => {
//   const service = testDataService();
//   return new Resource(service.app, {
//     service,
//     name: "my-resource",
//   });
// };

// /**
//  * A more complicated resource configuration
//  */
// export const complexService = () => {
//   const service = testDataService();

//   const person = new Resource(service.app, { service, name: "person" });

//   const club = new Resource(service.app, {
//     service,
//     name: "club",
//     comments: ["A club"],
//     attributeOptions: [{ name: "name" }],
//   });

//   person.addAttribute({
//     name: "first-name",
//     shortName: "fn",
//   });
//   person.addAttribute({
//     name: "last-name",
//     shortName: "ln",
//   });
//   person.addAttribute({
//     name: "height",
//     shortName: "ht",
//     type: ResourceAttributeType.INT,
//   });
//   person.addAttribute({
//     name: "is-frog",
//     shortName: "if",
//     type: ResourceAttributeType.BOOLEAN,
//     comments: ["Is this person a frog?"],
//   });
//   person.addAttribute({
//     name: "fav-colors",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: ResourceAttributeType.STRING,
//     comments: ["List opf favorite colors"],
//   });

//   // one specific club
//   person.addAttribute({
//     name: "one-club",
//     type: club,
//   });
//   // an array of clubs
//   person.addAttribute({
//     name: "array-of-clubs",
//     shortName: "ad",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: club,
//   });

//   // an array of persons
//   club.addAttribute({
//     name: "array-of-persons",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: person,
//   });

//   // operation
//   Operation.create(person);
//   Operation.read(person);
//   Operation.update(person);
//   Operation.delete(person);

//   return service;
// };

// /**
//  * Build a simple resource attribute to test with
//  */
// export const testResourceAttribute = () => {
//   const service = testDataService();
//   const resource = testResource();
//   return new ResourceAttribute(service.app, {
//     name: "my-attr",
//     resource,
//   });
// };

// /**
//  * Build a simple structure to test with
//  */
// export const testStructure = () => {
//   const service = testDataService();
//   return new Structure(service.app, { name: "my-structure" });
// };

// /**
//  * Build a simple operation to test with
//  */
// export const testOperation = () => {
//   const resource = testResource();
//   return new Operation(resource.service.app, {
//     name: "my-operation",
//     resource,
//   });
// };

// /**
//  * Build a nuxt site to test with
//  */
// export const testNuxtJsSite = () => {
//   return new NuxtJsSite({
//     parent: testFractureProject(),
//     name: "my-site",
//   });
// };
// export const testServiceTwoRelatedResources = () => {
//   const service = testDataService();
//   const person = new Resource(service.app, {
//     service,
//     name: "person",
//     attributeOptions: [{ name: "name", shortName: "n" }],
//   });
//   const club = new Resource(service.app, {
//     service,
//     name: "club",
//     attributeOptions: [{ name: "name", shortName: "n" }],
//   });
//   club.addAttribute({
//     name: "people",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: person,
//   });
//   person.addAttribute({
//     name: "clubs",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: club,
//   });
//   return { service, person, club };
// };
