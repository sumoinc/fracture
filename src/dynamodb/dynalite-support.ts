import { FractureComponent } from "../core/component";
import { FracturePackage } from "../core/fracture-package";
import { TypeScriptSource } from "../generators/ts/typescript-source";

export class DynaliteSupport extends FractureComponent {
  /**
   *
   * Writes a dynalite testable dynamo client to a typescript input file.
   *
   * @param file {TypeScriptSource}
   */
  static writeDynamoClient(file: TypeScriptSource, name: string = "dynamo") {
    file.comments([
      "Generate a DynamoDB client, configure it to use a local endpoint when needed",
      "to support unit testing with dynalite.",
      "",
      "https://www.npmjs.com/package/jest-dynalite",
    ]);
    file.open(`const config = {`);
    file.open(`...(process.env.MOCK_DYNAMODB_ENDPOINT && {`);
    file.line(`endpoint: process.env.MOCK_DYNAMODB_ENDPOINT,`);
    file.line(`sslEnabled: false,`);
    file.line(`region: 'local',`);
    file.close(`}),`);
    file.close(`}`);
    file.lines([
      `const client = new DynamoDBClient(config);`,
      `const ${name} = DynamoDBDocumentClient.from(client);`,
      ``,
    ]);
  }

  /**
   *
   * Writes imports required for dynalite support
   *
   * @param file {TypeScriptSource}
   */
  static writeJestImports(file: TypeScriptSource) {
    file.line(
      `import { createTables, deleteTables, startDb, stopDb, } from "jest-dynalite";`
    );
  }

  /**
   *
   * Writes a dynalite configuration used in jest tests
   *
   * @param file {TypeScriptSource}
   */
  static writeJestConfig(file: TypeScriptSource) {
    file.comments([
      "Sometimes dynalite tests can require a little additional",
      "time when you are running a lot of them in parallel.",
      "",
      "https://www.npmjs.com/package/jest-dynalite",
    ]);
    file.line("jest.setTimeout(10000);");
    file.line("");

    file.comments([
      "Starts and stops dynalite for each test",
      "",
      "https://www.npmjs.com/package/jest-dynalite",
    ]);
    file.line("beforeAll(startDb);");
    file.line("beforeEach(createTables);");
    file.line("afterEach(deleteTables);");
    file.line("afterAll(stopDb);");
    file.line("");
  }

  public readonly dynaliteConfig: TypeScriptSource;

  constructor(fracturePackage: FracturePackage) {
    super(fracturePackage);

    const { project } = this.fracturePackage;

    // add dynalite
    project.addDeps("jest-dynalite");
    // dynalite wants this peer fopr marshalling
    project.addPeerDeps("aws-sdk@^2.1336.0");

    this.dynaliteConfig = new TypeScriptSource(this, "jest-dynalite-config.js");

    // add dynalite setup to jest config
    const setupBeforeEnv = new TypeScriptSource(this, "setupBeforeEnv.ts");
    setupBeforeEnv.line(`import { setup } from "jest-dynalite";`);
    setupBeforeEnv.line("");
    setupBeforeEnv.line("// look for jest-dynalite-config.js in root");
    setupBeforeEnv.line(`setup(__dirname);`);
    setupBeforeEnv.line("");

    // add setup to the jest config
    project.jest!.addSetupFile("./setupBeforeEnv.ts");

    return this;
  }

  // build out the dynalite config
  preSynthesize() {
    const config = {
      tables: this.fracturePackage.services.map((service) => {
        const allGsi = service.dynamoTable.dynamoGsi.filter(
          (gsi) => gsi !== service.dynamoTable.keyDynamoGsi
        );
        return {
          TableName: service.dynamoTable.name,
          KeySchema: [
            {
              AttributeName: service.dynamoTable.pkName,
              KeyType: "HASH",
            },
            {
              AttributeName: service.dynamoTable.skName,
              KeyType: "RANGE",
            },
          ],
          AttributeDefinitions: service.dynamoTable.attrributeNames.map((s) => {
            return {
              AttributeName: s,
              AttributeType: "S",
            };
          }),
          GlobalSecondaryIndexes:
            allGsi.length > 0
              ? allGsi.map((gsi) => {
                  return {
                    IndexName: gsi.name,
                    KeySchema: [
                      {
                        AttributeName: gsi.pkName,
                        KeyType: "HASH",
                      },
                      {
                        AttributeName: gsi.skName,
                        KeyType: "RANGE",
                      },
                    ],
                    Projection: {
                      ProjectionType: "ALL",
                    },
                  };
                })
              : undefined,
          ProvisionedThroughput: {
            ReadCapacityUnits: 1,
            WriteCapacityUnits: 1,
          },
        };
      }),
      basePort: 8000 + this.fracturePackage.packageIndex * 100,
    };

    this.dynaliteConfig.line(
      `module.exports = ${JSON.stringify(config, null, 2)};`
    );

    this.dynaliteConfig.line("");
    super.preSynthesize();
  }
}
