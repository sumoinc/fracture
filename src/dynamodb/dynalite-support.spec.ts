import { addDynaliteSupport } from "./dynalite-support";
import { Fracture, FractureService } from "../core";
import { synthFile } from "../util/test-util";

let fracture: Fracture;

beforeEach(() => {
  fracture = new Fracture();
});

test("Smoke test", () => {
  const service = new FractureService(fracture, { name: "test" });
  addDynaliteSupport(service);

  const content = synthFile(fracture, "services/test/jest-dynalite-config.js");
  expect(content).toMatchSnapshot();
  console.log(content);
});
