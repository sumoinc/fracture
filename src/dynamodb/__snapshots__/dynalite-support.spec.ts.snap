// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Smoke test 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

module.exports = {
  \\"tables\\": [
    {
      \\"TableName\\": \\"test\\",
      \\"KeySchema\\": [
        {
          \\"AttributeName\\": \\"pk\\",
          \\"KeyType\\": \\"HASH\\"
        },
        {
          \\"AttributeName\\": \\"sk\\",
          \\"KeyType\\": \\"RANGE\\"
        }
      ],
      \\"AttributeDefinitions\\": [
        {
          \\"AttributeName\\": \\"pk\\",
          \\"AttributeType\\": \\"S\\"
        },
        {
          \\"AttributeName\\": \\"sk\\",
          \\"AttributeType\\": \\"S\\"
        },
        {
          \\"AttributeName\\": \\"idx\\",
          \\"AttributeType\\": \\"S\\"
        }
      ],
      \\"GlobalSecondaryIndexes\\": [
        {
          \\"IndexName\\": \\"loookup\\",
          \\"KeySchema\\": [
            {
              \\"AttributeName\\": \\"sk\\",
              \\"KeyType\\": \\"HASH\\"
            },
            {
              \\"AttributeName\\": \\"idx\\",
              \\"KeyType\\": \\"RANGE\\"
            }
          ],
          \\"Projection\\": {
            \\"ProjectionType\\": \\"ALL\\"
          }
        }
      ],
      \\"ProvisionedThroughput\\": {
        \\"ReadCapacityUnits\\": 1,
        \\"WriteCapacityUnits\\": 1
      }
    }
  ],
  \\"basePort\\": 7900
};
"
`;
